docs/libcudacxx/extended_api/work_stealing.rst

```diff
@@ -0,0 +1,111 @@
+
+Work stealing
+=============
+
+In header file ``<cuda/for_each_canceled>``:
+
+.. code:: cuda
+
+   namespace cuda {
+
+       template <int ThreadBlockRank = 3, typename UnaryFunction = ..unspecified..>
+       __device__ void for_each_canceled_block(UnaryFunction uf);
+
+   } // namespace cuda
+
+On devices with compute capability 10.0 or higher, it may leverage hardware acceleration.
+
+This API is mainly intended to implement work-stealing at thread-block level granularity.
+When compared against alternative work distribution techniques like `grid-stride loops <https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/>`__, which distribute load statically, or against other dynamic work distribution techniques using global memory concurrency, the main advantages of this API over these alternatives are:
+
+   - It performs work-stealing dynamically: thread blocks that finish work sooner may do more work than thread blocks whose work takes longer.
+   - It may cooperate with the GPU work-scheduler to respect work priorities and perform load-balancing.
+   - It may have lower work-stealing latency than global memory atomics.
+
+For better performance, extract the shared thread block prologue and epilog outside the lambda, and re-use it across thread-block iterations:
+
+  - Prologue: thread-block initialization code and data that is common to all thread blocks, e.g., ``__shared__`` memory allocation, their initialization, etc.
+  - Epilogue: thread-block finalization code that is common to all thread blocks, e.g., writing back shared memory to global memory, etc.
+
+**Mandates**:
+
+   - ``ThreadBlockRank`` equals the rank of the thread block: ``1``, ``2``, or ``3`` for one-dimensional, two-dimensional, and three-dimensional thread blocks, respectively.
+   - ``is_invokable_r_v<UnaryFunction, void, dim3>`` is true.
+
+**Preconditions**:
+
+   - All threads of the current thread-block shall call ``for_each_canceled_block`` **exactly once**.
+
+**Effects**:
+
+   - Invokes ``uf`` with ``blockIdx``, then repeatedly attempts to cancel the launch of another thread block in the current grid, and:
+
+      - on success, calls ``uf`` with that thread block's ``blockIdx`` and repeats,
+      - otherwise, it failed to cancel the launch of a thread block and it returns.
+
+Example
+-------
+
+This example shows how to perform work-stealing at thread-block granularity using this API.
+
+.. code:: cuda
+
+   // Before:
+
+   #include <cuda/math>
+   #include <cuda/for_each_canceled>
+   __global__ void vec_add(int* a, int* b, int* c, int n) {
+     // Extract common prologue outside the lambda, e.g.,
+     // - __shared__ or global (malloc) memory allocation
+     // - common initialization code
+     // - etc.
+
+     cuda::for_each_canceled_block<1>([=](dim3 block_idx) {
+       int idx = threadIdx.x + block_idx.x * blockDim.x;
+       // assert(block_idx == blockIdx); // May fail!
```

Comment:
Suggestion: I am not sure the assertion is the best way to transport that message. Maybe:
```suggestion
     cuda::for_each_canceled_block<1>([=](dim3 block_idx) {
       // block_idx may be different than the built-in blockIdx variable
       int idx = threadIdx.x + block_idx.x * blockDim.x;
```

Reply:
```suggestion
     cuda::for_each_canceled_block<1>([=](dim3 block_idx) {
       // block_idx may be different than the built-in blockIdx variable, that is:
       // assert(block_idx == blockIdx); // may fail!
       // so we need to use "block_idx" consistently inside for_each_canceled:
       int idx = threadIdx.x + block_idx.x * blockDim.x;
```

Reply:
LGTM!